packages/runtime/container-runtime/src/summary/runningSummarizer.ts

```diff
@@ -664,6 +661,139 @@ export class RunningSummarizer implements IDisposable {
+	) {
+		// Currently, we try 2 overall attempts with retries in each attempt. The second attempt will close the
+		// summarizer at the summarization call site but we need to keep this code for the time being. This will be
+		// reduced to a single attempt.
+		const attemptOptions: ISummarizeOptions[] = [
+			{ refreshLatestAck: false, fullTree: false },
+			{ refreshLatestAck: true, fullTree: false },
+		];
+		for (const summarizeOptions of attemptOptions) {
+			let previousStage: SummaryStage | undefined;
+
+			// The maximum number of retries across failure stages. If summarization does not succeed by these number
+			// of retries, quit and close the summarizer.
+			const maxRetries = defaultMaxRetries;
+			let currentRetries = -1;
+
+			// The maximum number of retries for a given stage. The default value is 1 but this can be set via failure
+			// params. Each failure is different and some can be succeeded by retrying more often than others.
+			// If summarization fails at different stages during retries, each stage gets some retries. However, the
+			// total number of retries across stages cannot exceed `maxRetries`.
+			let maxRetriesForStage = 1;
+			let currentRetriesForStage = -1;
+
+			type ShouldRetryResult =
+				| { retry: false }
+				| {
+						retry: true;
+						stage: SummaryStage;
+						delayMs: number;
+				  };
+			// Returns whether summarization should be retried based on the failure data.
+			const shouldRetry = (
+				failureData: ISubmitSummaryFailureResult | undefined,
+			): ShouldRetryResult => {
+				// For lastSummary, do not retry.
+				// If the failure data does not specify `retryAfterSeconds` or the stage at which summarization failed,
+				// do not retry.
+				// If maxRetries have been exceeded, do not retry anymore.
+				if (
+					reason === "lastSummary" ||
+					failureData === undefined ||
+					failureData.retryAfterSeconds === undefined ||
+					failureData.stage === undefined ||
+					currentRetries >= maxRetries
+				) {
+					return { retry: false };
+				}
+
+				// If summarization failed at the same stage as before, check if the max number of retries for this
+				// stage have been reached and if so, do not retry anymore.
+				if (failureData.stage === previousStage) {
+					if (currentRetriesForStage >= maxRetriesForStage) {
+						return { retry: false };
+					}
+					return {
+						retry: true,
+						stage: failureData.stage,
+						delayMs: failureData.retryAfterSeconds,
+					};
+				}
+
+				// If summarization failed at a different stage than before, reset the retry counters for this stage.
+				// Corner case - The first failure will be a different stage from previous stage (undefined) and in that
+				// case one try has already happened.
+				currentRetriesForStage = currentRetries === 0 ? 0 : -1;
+				// If the failure data specifies retryCount, use it. Else, use the default value.
+				maxRetriesForStage = failureData.retryCount ?? defaultMaxRetriesForStage;
```

Comment:
nit: I would prefer not to do any setting of variables here. It would be better to return them so I wouldn't have to recall that anything changed. Especially if we would want to isolate this function. Just a thought.

Reply:
That sounds reasonable. Let me see how much plumbing is needed to achieve that. It's all internal in one class method so its not a big deal anyway. Only reason I added this extra function is for clarity, I can inline it.

Reply:
Changed this to set variables at the call site rather than in this function.